File: BEmu/MarketDataRequest/MessageMarketSubscriptionCanceled.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="MarketDataRequest/MessageMarketSubscriptionCanceled.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.MarketDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class MessageMarketSubscriptionCanceled : Message
    {
        private readonly string _topicName;
        private readonly ElementMarketSubscriptionCancelReason _reason;

        internal MessageMarketSubscriptionCanceled(Subscription sub) : base(new Name("SubscriptionTerminated"), sub.CorrelationID, null)
        {
            this._topicName = sub.Security;
            this._reason = new ElementMarketSubscriptionCancelReason();
        }

        public override int NumElements { get { return 1; } }
        public override IEnumerable<Element> Elements { get { yield return this._reason; } }
        public override string TopicName { get { return this._topicName; } }

        public override object this[string name, int index] { get { return null; } }
        
        public override Element AsElement
        {
            get
            {
                return base.AsElement;
            }
        }

        public override string ToString()
        {
            StringBuilder result = new StringBuilder();
            result.AppendFormat("{0} = {{{1}", this.MessageType.ToString(), Environment.NewLine);
            result.Append(this._reason.PrettyPrint(1));
            result.AppendLine("}");

            return result.ToString();
        }
    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/MarketDataRequest/MessageMarketSubscriptionCanceled.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.MarketDataRequest.MessageMarketSubscriptionCanceled.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu.MarketDataRequest;

import com.bemu.BEmu.Message;
import com.bemu.BEmu.Subscription;
import com.bemu.BEmu.Name;
import com.bemu.BEmu.Element;

public class MessageMarketSubscriptionCanceled extends Message
{
    private final String _topicName;
    private final ElementMarketSubscriptionCancelReason _reason;

    MessageMarketSubscriptionCanceled(Subscription sub)
    {
    	super(new Name("SubscriptionTerminated"), sub.correlationID(), null);
    	
        this._topicName = sub.security();
        this._reason = new ElementMarketSubscriptionCancelReason();
    }
    
	public int numElements()
	{
		return 1;
	}
    
	public String topicName()
	{
		return this._topicName;
	}
	
	public Element getElement(String name) throws Exception
	{
		if(name.toUpperCase().equals(this._reason.name().toString().toUpperCase()))
			return this._reason;
			
		else
			return super.getElement(name);
	}
	
    public String toString()
    {
        StringBuilder result = new StringBuilder();
        result.append(String.format("{0} = {{{1}", this.messageType().toString(), System.getProperty("line.separator")));
        result.append(this._reason.prettyPrint(1));
        result.append("}" + System.getProperty("line.separator"));

        return result.toString();
    }

}
